File: kompartment/src/io/xml.js
/**
 * A small XML parser.
 *
 * Browsers have DOMParser and Node does not, and this project has no
 * dependencies, so the importer brings its own. It handles the subset that
 * Ecolego's writers actually emit -- elements, attributes, text, CDATA,
 * comments, the XML declaration and the five predefined entities -- and
 * nothing else. No namespaces, no DTDs, no entity declarations.
 *
 * Nodes are plain objects:
 *   { name, attrs: {..}, children: [node..], text: '<concatenated text>' }
 */

export class XMLError extends Error {
	constructor(message, position) {
		super(position != null ? `${message} (at character ${position})` : message);
		this.name = 'XMLError';
		this.position = position ?? null;
	}
}

// Null prototype: the name between `&` and `;` comes out of the file, and on a
// plain object `ENTITIES.constructor` is a function -- so `&constructor;` in a
// block's comment would have been replaced by the source of `Object`.
const ENTITIES = Object.assign(Object.create(null), {
	amp: '&', lt: '<', gt: '>', quot: '"', apos: "'",
});

// What ends a tag or attribute name. Hoisted: `readName` used to build this
// literal afresh for every character it looked at, which is every character
// of every name in the document.
const NAME_END = /[\s/>=]/;

export function decodeEntities(s) {
	if (!s.includes('&')) return s;
	return s.replace(/&(#x?[0-9a-fA-F]+|[a-zA-Z]+);/g, (whole, body) => {
		if (body[0] === '#') {
			const code = body[1] === 'x' || body[1] === 'X'
				? parseInt(body.slice(2), 16)
				: parseInt(body.slice(1), 10);
			// `String.fromCodePoint` throws a RangeError for anything that is
			// not a code point -- above U+10FFFF, or a surrogate half -- and a
			// file is entitled to contain nonsense. `&#1114112;` in one
			// comment killed the whole import with a message from deep inside
			// the string library; left as written, it is a comment that reads
			// oddly and nothing more.
			if (!Number.isInteger(code) || code < 0 || code > 0x10ffff) return whole;
			if (code >= 0xd800 && code <= 0xdfff) return whole;
			return String.fromCodePoint(code);
		}
		return ENTITIES[body] ?? whole;
	});
}

/**
 * @param {string} src
 * @returns {object} the root element node
 */
export function parseXML(src) {
	let i = 0;
	const n = src.length;

	const stack = [];
	let root = null;

	const fail = (msg) => { throw new XMLError(msg, i); };

	const skipSpace = () => {
		while (i < n && (src[i] === ' ' || src[i] === '\t' || src[i] === '\n' || src[i] === '\r')) i++;
	};

	const readName = () => {
		const start = i;
		while (i < n && !NAME_END.test(src[i])) i++;
		if (i === start) fail('Expected a name');
		return src.slice(start, i);
	};

	const addText = (raw) => {
		if (!stack.length) return;
		const parent = stack[stack.length - 1];
		parent.text += raw;
	};

	while (i < n) {
		const lt = src.indexOf('<', i);

		if (lt === -1) {
			addText(decodeEntities(src.slice(i)));
			break;
		}
		if (lt > i) addText(decodeEntities(src.slice(i, lt)));
		i = lt;

		// <!-- comment -->
		if (src.startsWith('<!--', i)) {
			const end = src.indexOf('-->', i + 4);
			if (end === -1) fail('Unterminated comment');
			i = end + 3;
			continue;
		}

		// <![CDATA[ ... ]]>  -- verbatim, no entity decoding
		if (src.startsWith('<![CDATA[', i)) {
			const end = src.indexOf(']]>', i + 9);
			if (end === -1) fail('Unterminated CDATA section');
			addText(src.slice(i + 9, end));
			i = end + 3;
			continue;
		}

		// <?xml ... ?>  and  <!DOCTYPE ...>
		if (src.startsWith('<?', i)) {
			const end = src.indexOf('?>', i + 2);
			if (end === -1) fail('Unterminated processing instruction');
			i = end + 2;
			continue;
		}
		if (src.startsWith('<!', i)) {
			const end = src.indexOf('>', i + 2);
			if (end === -1) fail('Unterminated declaration');
			i = end + 1;
			continue;
		}

		// </name>
		if (src.startsWith('</', i)) {
			i += 2;
			const name = readName();
			skipSpace();
			if (src[i] !== '>') fail(`Expected '>' closing </${name}`);
			i++;
			const open = stack.pop();
			if (!open) fail(`Unexpected closing tag </${name}>`);
			if (open.name !== name) {
				throw new XMLError(
					`Closing tag </${name}> does not match <${open.name}>`, i,
				);
			}
			continue;
		}

		// <name attr="value" ...>  or  <name .../>
		i++;
		const name = readName();
		// The attribute map has no prototype, for the same reason ENTITIES has
		// none: its keys are whatever the file wrote between `<name` and `>`,
		// and `attrs.constructor` on a plain object is a function.
		const node = { name, attrs: Object.create(null), children: [], text: '' };

		for (;;) {
			skipSpace();
			if (i >= n) fail('Unterminated tag');
			if (src[i] === '>') { i++; break; }
			if (src.startsWith('/>', i)) {
				i += 2;
				node.selfClosing = true;
				break;
			}
			const attrName = readName();
			skipSpace();
			if (src[i] !== '=') fail(`Expected '=' after attribute '${attrName}'`);
			i++;
			skipSpace();
			const quote = src[i];
			if (quote !== '"' && quote !== "'") {
				fail(`Expected a quoted value for attribute '${attrName}'`);
			}
			i++;
			const end = src.indexOf(quote, i);
			if (end === -1) fail(`Unterminated value for attribute '${attrName}'`);
			node.attrs[attrName] = decodeEntities(src.slice(i, end));
			i = end + 1;
		}

		if (stack.length) {
			stack[stack.length - 1].children.push(node);
		} else if (root) {
			fail('More than one root element');
		} else {
			root = node;
		}
		if (!node.selfClosing) stack.push(node);
	}

	if (stack.length) {
		throw new XMLError(`Unclosed element <${stack[stack.length - 1].name}>`);
	}
	if (!root) throw new XMLError('The document has no elements');
	return root;
}

// --- convenience accessors -------------------------------------------------

/** First direct child with the given name, or null. */
export function child(node, name) {
	if (!node) return null;
	return node.children.find((c) => c.name === name) ?? null;
}

/** All direct children with the given name. */
export function children(node, name) {
	if (!node) return [];
	return node.children.filter((c) => c.name === name);
}

/** Text of a direct child, trimmed, or a fallback. */
export function childText(node, name, fallback = null) {
	const c = child(node, name);
	if (!c) return fallback;
	const t = c.text.trim();
	return t === '' ? fallback : t;
}

export function childNumber(node, name, fallback = null) {
	const t = childText(node, name);
	if (t == null) return fallback;
	const v = Number(t);
	return Number.isFinite(v) ? v : fallback;
}

export function childBool(node, name, fallback = null) {
	const t = childText(node, name);
	if (t == null) return fallback;
	return t.toLowerCase() === 'true';
}

/**
 * Depth-first walk over every element.
 *
 * Iterative, with its own stack. `yield* walk(c)` recursed once per level of
 * nesting *and* once per level for every value yielded through it, so five
 * thousand nested elements -- a few tens of kilobytes of XML, well inside what
 * a file can carry -- overflowed the call stack. The parser itself has always
 * been iterative; this was the half that was not, and `find` below is what
 * `eco.js` reads a document with.
 */
export function* walk(node) {
	const stack = [node];
	while (stack.length) {
		const el = stack.pop();
		yield el;
		// Pushed in reverse so siblings come out in document order.
		const kids = el.children;
		for (let i = kids.length - 1; i >= 0; i--) stack.push(kids[i]);
	}
}

/** First element anywhere under `node` with the given name. */
export function find(node, name) {
	for (const el of walk(node)) {
		if (el.name === name) return el;
	}
	return null;
}

File: kompartment/python/kompartment/importers/_xml.py
"""A small XML reader, the one the Ecolego importer reads model.xml with.

A port of ``src/io/xml.js``, kept rather than replaced by the standard
library's parsers so that a file is read the same way here as in the
application: the same subset of XML is accepted -- elements, attributes, text,
CDATA, comments, the XML declaration and the five predefined entities, and
nothing else (no namespaces, no DTDs, no entity declarations) -- the same
malformed files are refused with the same message, and the text of an element
is exactly what the application sees. ``xml.etree`` would read a namespace
prefix, a DTD entity or a stray ``&nbsp;`` differently, and refuse files this
reads.

An element is a :class:`Node`: its ``name``, its ``attrs``, its ``children``
and its ``text`` -- every piece of text directly inside it, concatenated.
"""

from __future__ import annotations

import math
import re
from typing import Dict, Iterator, List, Optional

from ._eco_maps import JS_SPACE_CLASS, js_trim, to_number, utf16_index


class XMLError(ValueError):
    """A document this reader cannot read. ``position`` is where it gave up,
    counted in UTF-16 code units as the application counts, or ``None``."""

    def __init__(self, message: str, position: Optional[int] = None) -> None:
        super().__init__(f'{message} (at character {position})' if position is not None else message)
        self.position = position


class Node:
    """One element: ``name``, ``attrs`` (in the order written; a repeated
    attribute keeps its first place and its last value), ``children`` and
    ``text``."""

    __slots__ = ('name', 'attrs', 'children', 'text', 'self_closing')

    def __init__(self, name: str) -> None:
        self.name = name
        self.attrs: Dict[str, str] = {}
        self.children: List['Node'] = []
        self.text = ''
        self.self_closing = False

    def __repr__(self) -> str:
        return f'<Node {self.name} {len(self.children)} children>'


ENTITIES = {'amp': '&', 'lt': '<', 'gt': '>', 'quot': '"', 'apos': "'"}

_ENTITY = re.compile(r'&(#x?[0-9a-fA-F]+|[a-zA-Z]+);')
# What ends a tag or attribute name: JavaScript's white space, `/`, `>` or `=`.
_NAME = re.compile('[^' + JS_SPACE_CLASS + '/>=]*')
_SPACE = re.compile('[ \t\n\r]*')
_DIGITS = {10: re.compile('[0-9]*'), 16: re.compile('[0-9a-fA-F]*')}


def _parse_int(text: str, radix: int) -> Optional[int]:
    # `parseInt`: the longest run of digits at the start, or NaN (None).
    run = _DIGITS[radix].match(text).group(0)
    return int(run, radix) if run else None


def _entity(m: 're.Match[str]') -> str:
    whole, body = m.group(0), m.group(1)
    if body[0] == '#':
        code = _parse_int(body[2:], 16) if body[1:2] in ('x', 'X') else _parse_int(body[1:], 10)
        # Not a code point, or half of a surrogate pair: left as written, as a
        # comment that reads oddly rather than an import that fails.
        if code is None or code < 0 or code > 0x10FFFF:
            return whole
        if 0xD800 <= code <= 0xDFFF:
            return whole
        return chr(code)
    return ENTITIES.get(body, whole)


def decode_entities(text: str) -> str:
    """The five predefined entities and numeric character references decoded;
    anything else (``&nbsp;``, ``&#1114112;``) left as written."""
    if '&' not in text:
        return text
    return _ENTITY.sub(_entity, text)


def parse_xml(src: str) -> Node:
    """Reads a document; returns its root element. Raises :class:`XMLError`
    for what it cannot read, with the application's message."""
    i = 0
    n = len(src)
    stack: List[Node] = []
    parts: List[List[str]] = []  # the text of each open element, in pieces
    root: Optional[Node] = None

    def fail(message: str) -> None:
        raise XMLError(message, utf16_index(src, i))

    def read_name() -> str:
        nonlocal i
        start = i
        i = _NAME.match(src, i).end()
        if i == start:
            fail('Expected a name')
        return src[start:i]

    def skip_space() -> None:
        nonlocal i
        i = _SPACE.match(src, i).end()

    def add_text(raw: str) -> None:
        if parts:
            parts[-1].append(raw)

    while i < n:
        lt = src.find('<', i)
        if lt == -1:
            add_text(decode_entities(src[i:]))
            break
        if lt > i:
            add_text(decode_entities(src[i:lt]))
        i = lt

        if src.startswith('<!--', i):
            end = src.find('-->', i + 4)
            if end == -1:
                fail('Unterminated comment')
            i = end + 3
            continue
        # CDATA is verbatim: no entity decoding.
        if src.startswith('<![CDATA[', i):
            end = src.find(']]>', i + 9)
            if end == -1:
                fail('Unterminated CDATA section')
            add_text(src[i + 9:end])
            i = end + 3
            continue
        if src.startswith('<?', i):
            end = src.find('?>', i + 2)
            if end == -1:
                fail('Unterminated processing instruction')
            i = end + 2
            continue
        if src.startswith('<!', i):
            end = src.find('>', i + 2)
            if end == -1:
                fail('Unterminated declaration')
            i = end + 1
            continue

        if src.startswith('</', i):
            i += 2
            name = read_name()
            skip_space()
            if i >= n or src[i] != '>':
                fail(f"Expected '>' closing </{name}")
            i += 1
            if not stack:
                fail(f'Unexpected closing tag </{name}>')
            open_ = stack.pop()
            open_.text = ''.join(parts.pop())
            if open_.name != name:
                raise XMLError(f'Closing tag </{name}> does not match <{open_.name}>',
                               utf16_index(src, i))
            continue

        i += 1
        name = read_name()
        node = Node(name)
        while True:
            skip_space()
            if i >= n:
                fail('Unterminated tag')
            if src[i] == '>':
                i += 1
                break
            if src.startswith('/>', i):
                i += 2
                node.self_closing = True
                break
            attr = read_name()
            skip_space()
            if i >= n or src[i] != '=':
                fail(f"Expected '=' after attribute '{attr}'")
            i += 1
            skip_space()
            quote = src[i] if i < n else ''
            if quote not in ('"', "'"):
                fail(f"Expected a quoted value for attribute '{attr}'")
            i += 1
            end = src.find(quote, i)
            if end == -1:
                fail(f"Unterminated value for attribute '{attr}'")
            node.attrs[attr] = decode_entities(src[i:end])
            i = end + 1

        if stack:
            stack[-1].children.append(node)
        elif root is not None:
            fail('More than one root element')
        else:
            root = node
        if not node.self_closing:
            stack.append(node)
            parts.append([])

    if stack:
        raise XMLError(f'Unclosed element <{stack[-1].name}>')
    if root is None:
        raise XMLError('The document has no elements')
    return root


# --- convenience accessors ----------------------------------------------------------

def child(node: Optional[Node], name: str) -> Optional[Node]:
    """The first direct child called ``name``, or ``None``."""
    if node is None:
        return None
    for c in node.children:
        if c.name == name:
            return c
    return None


def children(node: Optional[Node], name: str) -> List[Node]:
    """Every direct child called ``name``."""
    if node is None:
        return []
    return [c for c in node.children if c.name == name]


def child_text(node: Optional[Node], name: str, fallback: Optional[str] = None) -> Optional[str]:
    """The text of the first child called ``name``, trimmed, or ``fallback``
    when there is no such child or its text is blank."""
    c = child(node, name)
    if c is None:
        return fallback
    t = js_trim(c.text)
    return fallback if t == '' else t


def child_number(node: Optional[Node], name: str, fallback: Optional[float] = None) -> Optional[float]:
    """A child's text as a finite number, or ``fallback``."""
    t = child_text(node, name)
    if t is None:
        return fallback
    v = to_number(t)
    return v if math.isfinite(v) else fallback


def child_bool(node: Optional[Node], name: str, fallback: Optional[bool] = None) -> Optional[bool]:
    """Whether a child's text is ``true`` (in any case); ``fallback`` when
    there is none."""
    t = child_text(node, name)
    if t is None:
        return fallback
    return t.lower() == 'true'


def walk(node: Node) -> Iterator[Node]:
    """Every element under ``node``, ``node`` first, depth first in document
    order. Iterative, so nesting depth is no limit."""
    stack = [node]
    while stack:
        el = stack.pop()
        yield el
        stack.extend(reversed(el.children))


def find(node: Node, name: str) -> Optional[Node]:
    """The first element anywhere under ``node`` (``node`` included) called
    ``name``, or ``None``."""
    for el in walk(node):
        if el.name == name:
            return el
    return None
